### PR TITLE
Add lifetime to datasets

### LINKED ISSUE #10746: Add a lifetime for a given dataset shipping to disk sites from Tier0
**Impact of the new feature**
Inform which systems would be affected by this new feature:
The systems and components that would be affected by this new future is the WMAgent with the following components

- DBS3Buffer
- RucioInjector

**Is your feature request related to a problem? Please describe.**
We should not manage disk space in general due that disk space is not infinity for CMS. When Tier0 create new datasets and those ships to disk sites, the data will remain there unless a service account with enough privilegies delete the rule that is protecting that dataset. With this feature Tier0 gives that attribute to a dataset that ships to disk

**Describe the solution you'd like**
For a given dataset produced by Tier0 and that ships to disk sites assign a lifetime.

**Describe alternatives you've considered**
I considered not to involve the DBS3Buffer for this development giving to all the datasets the same lifetime adding one more argument within RucioInjectorPoller.py as follows
```
            else:
                 # then it's a T0 container placement
                 if not rseName.endswith("_Tape"):
                     #Assign lifetime to container rules shipping to Disk sites
                     ruleKwargs['lifetime'] = 18*30*24*60*60
```
But this approach was wrong due that Tier0 needs the lifetime configurable for a given dataset and given as an argument from our Configuration file. Thats why this development goes attached to [getUnsubscribedDsets](https://github.com/dmwm/WMCore/blob/125f46fc753d406b83c9061d694ba60ba7e68dcd/src/python/WMComponent/RucioInjector/RucioInjectorPoller.py#L429) 
**Additional context**
Tested with a replay and generated the following 3  container rules with expiration date. With the replay we assigned a default value of 18 months, but is configurable in function of the dataset
```
Id:                         78abfffc8fe248a7ad64a3bac43f5b27
Account:                    tier0_replay
Scope:                      cms
Name:                       /Cosmics/Tier0_REPLAY_2021-v2108110537/RAW
RSE Expression:             T0_CH_CERN_Disk
Copies:                     1
State:                      OK
Locks OK/REPLICATING/STUCK: 97/0/0
Grouping:                   ALL
Expires at:                 2023-08-01 05:02:08
Locked:                     False
Weight:                     None
Created at:                 2021-08-11 05:02:08
Updated at:                 2021-08-11 07:25:55
Error:                      None
Subscription Id:            None
Source replica expression:  None
Activity:                   T0 Export
Comment:                    T0 WMAgent automatic container rule
Ignore Quota:               False
Ignore Availability:        False
Purge replicas:             False
Notification:               NO
End of life:                None
Child Rule Id:              None

Id:                         2a73a72836ef4221bc2d6d91775b50e8
Account:                    tier0_replay
Scope:                      cms
Name:                       /StreamExpressCosmics/Tier0_REPLAY_2021-PromptCalibProdSiPixel-Express-v2108110537/ALCAPROMPT
RSE Expression:             T0_CH_CERN_Disk
Copies:                     1
State:                      OK
Locks OK/REPLICATING/STUCK: 132/0/0
Grouping:                   ALL
Expires at:                 2022-08-06 05:07:26
Locked:                     False
Weight:                     None
Created at:                 2021-08-11 05:07:26
Updated at:                 2021-08-11 06:50:29
Error:                      None
Subscription Id:            None
Source replica expression:  None
Activity:                   T0 Export
Comment:                    T0 WMAgent automatic container rule
Ignore Quota:               False
Ignore Availability:        False
Purge replicas:             False
Notification:               NO
End of life:                None
Child Rule Id:              None

Id:                         5d3a6c335df347778fa4144dac17dc4e
Account:                    tier0_replay
Scope:                      cms
Name:                       /MinimumBias/Tier0_REPLAY_2021-v2108110537/RAW
RSE Expression:             T0_CH_CERN_Disk
Copies:                     1
State:                      OK
Locks OK/REPLICATING/STUCK: 29/0/0
Grouping:                   ALL
Expires at:                 2023-02-02 06:04:09
Locked:                     False
Weight:                     None
Created at:                 2021-08-11 06:04:09
Updated at:                 2021-08-11 08:02:48
Error:                      None
Subscription Id:            None
Source replica expression:  None
Activity:                   T0 Export
Comment:                    T0 WMAgent automatic container rule
Ignore Quota:               False
Ignore Availability:        False
Purge replicas:             False
Notification:               NO
End of life:                None
Child Rule Id:              None
```

### PR DESCRIPTION
**Pull request by jhonatanamado**
Add lifetime to dataset shipping to disk sites

Fixes #10746 

#### Status
Tested with a replay

#### Description
For a given dataset  processed by Tier0 and shipping to disk sites, add a new attribute of lifetime to the rucio container rule.

#### Is it backward compatible (if not, which system it affects?)
MAYBE. Needs to be properly evaluated due that this PR creates a new column in the oracle/mysql database.

#### Related PRs
The following PR in dmwm/T0 [#4601](https://github.com/dmwm/T0/pull/4601)

#### External dependencies / deployment changes
No

### CODE REVIEW

**Review by amaltaro — CHANGES_REQUESTED** [review]
@jhonatanamado thanks for proposing these changes, Jhonatan.
I have made a few comments along the code that need attention. Since it's going to be a new workflow parameter, we should make sure it's supported in the StdBase module as well:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/StdSpecs/StdBase.py#L1141

and properly validated.

This pull request is in the right direction, but I think I'd like to first give some thought to this issue https://github.com/dmwm/WMCore/issues/9639, properly discuss and redesign it considering the new Data Management tool that we have today.

File: src/python/WMComponent/DBS3Buffer/MySQL/Create.py [hunk 1]
```diff
@@ -49,8 +49,9 @@ def __init__(self, logger=None, dbi=None, params=None):
                  subscribed             INTEGER      DEFAULT 0,
                  phedex_group           VARCHAR(100),
                  delete_blocks          INTEGER,
+                 dataset_lifetime           INTEGER      DEFAULT 0,
                  PRIMARY KEY (id),
-                 CONSTRAINT uq_dbs_dat_sub UNIQUE (dataset_id, site, custodial, auto_approve, move, priority))"""
+                 CONSTRAINT uq_dbs_dat_sub UNIQUE (dataset_id, site, custodial, auto_approve, move, priority, dataset_lifetime))"""
```
Comment (amaltaro) [inline src/python/WMComponent/DBS3Buffer/MySQL/Create.py:54]: Do we need to make dataset_lifetime part of the table unique constraint? I'm inclined to say no, but I might be missing something.

File: src/python/WMCore/WMSpec/WMWorkload.py [hunk 2]
```diff
@@ -1573,6 +1573,7 @@ def getSubscriptionInformation(self, initialTask=None):
                     subInfo[dataset]["NonCustodialGroup"] = solveGroupConflicts(
                         taskSubInfo[dataset]["NonCustodialGroup"],
                         subInfo[dataset]["NonCustodialGroup"])
+                    subInfo[dataset]["Dataset_lifetime"] = solveGroupConflicts(taskSubInfo[dataset]["Dataset_lifetime"], subInfo[dataset]["Dataset_lifetime"])
```
Comment (amaltaro) [inline src/python/WMCore/WMSpec/WMWorkload.py:1576]: the key name here should be `DatasetLifetimeDisk `, as defined in the StdBase spec.

File: src/python/WMCore/WMSpec/WMWorkload.py [hunk 3]
```diff
@@ -1483,7 +1483,7 @@ def setSubscriptionInformation(self, initialTask=None, custodialSites=None,
                                    custodialGroup="DataOps", nonCustodialGroup="DataOps",
                                    priority="Low", primaryDataset=None,
                                    useSkim=False, isSkim=False,
-                                   dataTier=None, deleteFromSource=False):
+                                   dataTier=None, deleteFromSource=False, dataset_lifetime=None):
```
Comment (amaltaro) [inline src/python/WMCore/WMSpec/WMWorkload.py:1486]: Can you please keep the same variable name convention defined around this code (lower camel case), thus `datasetLifetime`?

File: src/python/WMCore/WMSpec/WMTask.py [hunk 4]
```diff
@@ -1082,6 +1083,7 @@ def getSubscriptionInformation(self):
                                        "NonCustodialSites": outputSection.nonCustodialSites,
                                        "AutoApproveSites": outputSection.autoApproveSites,
                                        "Priority": outputSection.priority,
+                                       "Dataset_lifetime": outputSection.dataset_lifetime,
```
Comment (amaltaro) [inline src/python/WMCore/WMSpec/WMTask.py:1086]: This needs to be updated according to the StdBase spec schema as well.

File: src/python/WMComponent/RucioInjector/RucioInjectorPoller.py [hunk 5]
```diff
@@ -465,6 +466,9 @@ def insertContainerRules(self):
             else:
                 # then it's a T0 container placement
                 ruleKwargs['priority'] = 4
+                if not rseName.endswith("_Tape"):
```
Comment (amaltaro) [inline src/python/WMComponent/RucioInjector/RucioInjectorPoller.py:469]: should we also make sure that the `lifetime` is greater than 0 (thus, not using the database default value)?

**Review by amaltaro — CHANGES_REQUESTED** [review]
Thanks for providing further changes to this PR, Jhonatan.
I have two generic comments to make:
1) there are different places where this feature is defined, and they require different naming conventions (and some might require your attention):
a) workflow spec file is correctly defined as `DatasetLifetimeDisk`
b) variable name used in the project is lower camel case. So please replace  `dataset_lifetime` by `datasetLifetime`
c) table variable name is better to be made in case insensitive way, so your `dataset_lifetime` looks correct to me.
2) regarding this feature itself, it's based on an old feature that is soon going to be changed, since it's still based on the PhEDEx era. What I would like is to actually refactor this code and make it meaningful to our current Data Management system. We can merge it without that refactoring, but I just wanted to say that further changes will be required on the T0 once we start refactoring this code.
Thanks!

**Comment by jhonatanamado** [comment]
Thanks Alan, 
I will follow all your suggestions regarding the first bullet. And indeed I will refactor this code and make it meaningful to the current Data Management system. Probably is the best way to go despite that means some changes on T0 code.

File: src/python/WMComponent/RucioInjector/RucioInjectorPoller.py (hunk 5 above)
Comment (jhonatanamado) [reply src/python/WMComponent/RucioInjector/RucioInjectorPoller.py:469]: Hi Alan I don't know if I understood your suggestion but this is solved from T0 code in this [line](https://github.com/dmwm/T0/blob/15ab4669bb6830dfc9043b810e33a72657ec7ea0/src/python/T0/RunConfig/Tier0Config.py#L551) with a default value.

**Pull request closed without merging**